lib/apollo/client.js

```diff
@@ -0,0 +1,152 @@
+      if (PageComponent.getInitialProps) {
+        pageProps = await PageComponent.getInitialProps(ctx);
+      }
+
+      // Only on the server:
+      if (typeof window === "undefined") {
+        // When redirecting, the response is finished.
+        // No point in continuing to render
+        if (ctx.res && ctx.res.finished) {
+          return pageProps;
+        }
+
+        // Only if ssr is enabled
+        if (ssr) {
+          try {
+            // Run all GraphQL queries
+            const { getDataFromTree } = await import("@apollo/react-ssr");
+            await getDataFromTree(
+              <AppTree
+                pageProps={{
+                  ...pageProps,
+                  apolloClient,
+                }}
+              />
+            );
+          } catch (error) {
+            // Prevent Apollo Client GraphQL errors from crashing SSR.
+            // Handle them in components via the data.error prop:
+            // https://www.apollographql.com/docs/react/api/react-apollo.html#graphql-query-data-error
+            console.error("Error while running `getDataFromTree`", error);
+          }
+
+          // getDataFromTree does not call componentWillUnmount
+          // head side effect therefore need to be cleared manually
+          Head.rewind();
+        }
+      }
+
+      // Extract query data from the Apollo store
+      const apolloState = apolloClient.cache.extract();
+
+      return {
+        ...pageProps,
+        apolloState,
+      };
+    };
+  }
+
+  return WithApollo;
+}
+
+/**
+ * Always creates a new apollo client on the server
+ * Creates or reuses apollo client in the browser.
+ * @param  {Object} initialState
+ */
+function initApolloClient(ctx, initialState) {
+  // Make sure to create a new client for every server-side request so that data
+  // isn't shared between connections (which would be bad)
+  if (typeof window === "undefined") {
+    return createApolloClient(ctx, initialState);
+  }
+
+  // Reuse client on the client-side
+  if (!globalApolloClient) {
+    globalApolloClient = createApolloClient(ctx, initialState);
+  }
+
+  return globalApolloClient;
+}
+
+/**
+ * Creates and configures the ApolloClient
+ * @param  {Object} [initialState={}]
+ */
+function createApolloClient(ctx = {}, initialState = {}) {
+  const ssrMode = typeof window === "undefined";
+  const cache = new InMemoryCache().restore(initialState);
+
+  // Check out https://github.com/zeit/next.js/pull/4611 if you want to use the AWSAppSyncClient
+  return new ApolloClient({
+    ssrMode,
+    link: createIsomorphLink(ctx),
+    cache,
+  });
+}
+
+function createIsomorphLink(ctx) {
+  if (typeof window === "undefined") {
+    const { SchemaLink } = require("apollo-link-schema");
+    const { schema } = require("./schema");
+    return new SchemaLink({ schema, context: ctx });
+  } else {
+    const { HttpLink } = require("apollo-link-http");
+
+    return new HttpLink({
+      uri: "/graphql",
```

Comment:
According to this article this should be an absolute URL in order for SSR to work

https://dev.to/givehug/next-js-apollo-client-and-server-on-a-single-express-app-55l6